### PR TITLE
feat(sprintf): support Go-compatible %q for strings

### PR DESCRIPTION
**Pull request by vitaliytv**
## Summary

Add support for Go-syntax `%q` formatting for Regorus string values. The implementation now matches Go 1.27.1 / OPA 1.20.2 for the supported string forms:

- emits a double-quoted Go string literal;
- uses Go's short escapes and `\\xNN` / `\\uNNNN` / `\\UNNNNNNNN` escapes;
- preserves printable non-ASCII characters and does not HTML-escape `<`, `>`, or `&`;
- applies the width and precision forms already recognized by the current Regorus `sprintf` parser (`%Nq`, `%0Nq`, and `%.Nq`);
- appends directly to the `sprintf` output buffer and propagates allocator memory-limit errors.

## Correctness changes

The initial implementation approximated Go's `strconv.IsPrint` with Rust character predicates. That differs for Unicode format characters, private-use characters, noncharacters, and unassigned code points. This revision ports the compact generated tables and lookup algorithm from Go 1.27.1's `strconv/isprint.go` (Unicode 17.0.0), so values such as U+00AD, U+200B, U+E000, U+FDD0, and U+0378 are escaped exactly like Go.

Go's `%q` applies string precision before quoting and field width after quoting, both measured in Unicode code points. The supported Regorus forms now do the same, including Unicode input and zero padding.

## Verification

Compared directly with OPA 1.20.2 (built with Go 1.27.1) for regular strings, every escape class, printable Unicode, the Unicode edge cases above, width, zero padding, precision, and emoji width/precision.

Local checks:

- [x] `cargo test --locked`
- [x] `cargo test --locked --no-default-features --lib`
- [x] `cargo check --locked --no-default-features`
- [x] `cargo xtask test-no-std`
- [x] allocator memory-limit regression test for `%q`
- [x] `cargo clippy --locked --lib --all-features -- -D warnings`
- [x] repository pre-commit hook (all build, fmt, and clippy stages)
- [x] repository pre-push hook

## Scope and full Go `%q`

This PR intentionally handles `Value::String` only and preserves the repository's existing partial `sprintf` parser. A complete implementation of Go's `%q` would also require:

- combined width and precision (`%8.3q`);
- `+` for ASCII-only quoting, `#` for backquoted strings when allowed, and `-` for right padding;
- dynamic width/precision and argument indexes (`*`, `[n]`);
- integer/rune formatting as a single-quoted character literal.

OPA adds another compatibility layer before calling `fmt.Sprintf`: strings remain Go strings, fitting integers become Go integers, while booleans, nulls, arrays, objects, and sets are converted to their OPA string representation. Reproducing all of that is broader than a quoting helper and should be implemented with a proper format-spec parser and explicit OPA value-conversion rules rather than incrementally extending this string-only branch.

### CODE REVIEW

**Comment by vitaliytv** [comment]
Follow-up for the complete format-spec parser and broader Go/OPA `%q` semantics: #810. It is intentionally stacked on this PR.

**Comment by vitaliytv** [comment]
Would you prefer this change to be reviewed in two stages, or as one larger PR?

1. Keep #809 focused on correct Go-compatible string quoting (including width and precision), and review #810 separately for the complete format-spec parser, flags, dynamic/indexed operands, and broader OPA value semantics. This is my preference because it keeps the core fix smaller and makes the additional compatibility scope explicit.
2. Fold the #810 commit into #809 and close the follow-up, if a single complete implementation would be easier to review.

#810 is currently stacked on this PR, so its diff against `main` temporarily includes #809. I am happy to reorganize the commits according to whichever approach you prefer.

**Comment by anakrish** [comment]
@vitaliytv Thank you for the contribution. Going to take a deeper look at this today/tomorrow.

**Comment by anakrish** [comment]
Hey @vitaliytv, Overall looks good to me.

Can you also lock down the following as a test case:

```yaml
  - note: "%q must escape Unicode line and paragraph separators"
    data: {}
    modules:
      - |
        package test

        line_separator := sprintf("%q", ["a\u2028b"])
        paragraph_separator := sprintf("%q", ["a\u2029b"])
    query: data.test
    want_result:
      line_separator: "\"a\\u2028b\""
      paragraph_separator: "\"a\\u2029b\""
```

**Comment by vitaliytv** [comment]
> Hey Vitalii Tverdokhlib (@vitaliytv), Overall looks good to me.
> 
> Can you also lock down the following as a test case:
> 
> ```yaml
>   - note: "%q must escape Unicode line and paragraph separators"
>     data: {}
>     modules:
>       - |
>         package test
> 
>         line_separator := sprintf("%q", ["a\u2028b"])
>         paragraph_separator := sprintf("%q", ["a\u2029b"])
>     query: data.test
>     want_result:
>       line_separator: "\"a\\u2028b\""
>       paragraph_separator: "\"a\\u2029b\""
> ```

done

File: src/builtins/strings.rs [hunk 1]
```diff
@@ -407,6 +484,8 @@ fn sprintf(span: &Span, params: &[Ref<Expr>], args: &[Value], _strict: bool) ->
                 bail!(args_span.error(&format!("number specified for format verb {verb}.")));
             }
 
+            ('q', Value::String(sv)) => append_go_quoted(&mut s, sv.as_ref(), width)?,
```
[the remark this answers is not in the document]
Comment (vitaliytv) [reply src/builtins/strings.rs:487]: Fixed in https://github.com/microsoft/regorus/commit/fd939063036fd0f202766240d804f77763ddbe16: added `tests/rvm/rego/cases/sprintf.yaml`, exercising `%q` quoting plus Unicode width and precision through RVM compilation and builtin dispatch.

File: src/builtins/strings/go_is_print.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
+// Copyright 2013 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Adapted from Go 1.27.1's generated strconv/isprint.go tables (Unicode 17.0.0).
+// Keep this data synchronized with the Go version used as the OPA compatibility
+// reference. The four tables below occupy 3,560 bytes.
```
[the remark this answers is not in the document]
Comment (vitaliytv) [reply src/builtins/strings/go_is_print.rs]: Corrected in https://github.com/microsoft/regorus/commit/02695ea8612d1ffa4b33e3b8722a92b9cfaf5dad. Verified against the Go 1.27.1 generated source and copied arrays: isNotPrint32 has 121 u16 entries, so the exact footprint is 3,594 bytes, not 3,592.

**Review by anakrish — APPROVED** [review]
